### ISSUE TITLE
APE playback freezes cmus

### ISSUE THREAD

**Issue by shoober420** [cmus/cmus#812]
---

1) `uname -a`:
```
Linux _QzWxEc_-717 4.17.9-1-zen #1 ZEN SMP PREEMPT Sun Jul 22 20:23:46 UTC 2018 x86_64 GNU/Linux
```

2) `cmus --version`:
```
cmus v2.8.0-rc0-90-g17bf542
```

3) `cmus --plugins`:
```
Input Plugins: /usr/lib/cmus/ip
  mad:
    Priority: 55
    File Types: mp3 mp2
    MIME Types: audio/mpeg audio/x-mp3 audio/x-mpeg
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp
  mpc:
    Priority: 50
    File Types: mpc mpp mp+
    MIME Types: audio/x-musepack
  opus:
    Priority: 50
    File Types: opus
    MIME Types:
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  modplug:
    Priority: 50
    File Types: mod s3m xm it 669 amf ams dbm dmf dsm far mdl med mtm okt ptm stm ult umx mt2 psm
    MIME Types:
  wavpack:
    Priority: 50
    File Types: wv
    MIME Types: audio/x-wavpack
  mp4:
    Priority: 50
    File Types: mp4 m4a m4b
    MIME Types:
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  cdio:
    Priority: 50
    File Types:
    MIME Types: x-content/audio-cdda
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg shn tak tta wav webm wma wv
    MIME Types:

Output Plugins: /usr/lib/cmus/op
  pulse
  alsa
  oss
  jack
  ao
```

After the first track finishes from an APE file in the playlist, cmus locks up. FLAC playback works perfectly fine. This only happens with APE files. The APE cuesheet is being used to create the tracklist and play via the "Library view (1)".

ffmpeg 1:4.0.2-1 on Arch Linux

**Comment by shoober420** [comment]
Wait never mind, this still happens. It sucks because I then have to Ctr+/ the program which is frozen, then reopen it and add it to the playlist again after every song. I’m currently running cmus outside of X. 

Then after I quit cmus, every time I need to sudo, it locks up after I put in my password. So then I have to Ctrl+/ the sudo password prompt, start X, then run sudo commands from there or change tty from the terminal.   

REEEEEEEEEEEEEEEEEEEEEEEEEEEEE

Ishdhejdifukdjdjdhdufifjejejdjjendjfjdjjdjdjdufjdjdjdijf

**Comment by shoober420** [comment]
Changing "buffer_seconds" and "dsp.ao.buffer.size" values from higher to lower doesn't remedy the issue. APE playback still freezes cmus.

**Comment by nefthy** [comment]
Could you please compile a clear list of actions to reproduce this issue?

**Comment by shoober420** [comment]
I just compiled a fresh version of cmus from the latest head. If I play an .APE file from the browser view (5), cmus will not lock up. But when added the directory that the .APE file is located in, and going to the playlist view (1), and cmus using the .CUE file to play the .APE file, that’s when cmus locks up. I believe it has something to do with cue handling of APE files. Numerous APE cue files were tested.

**Comment by shoober420** [comment]
Just did a fresh compile of cmus using the latest version of ffmpeg, and APE playback still freezes cmus.

**Comment by gdrbyKo1** [comment]
Same issue here with cmus 2748d40bb670558b523d5b47b4af442e82c7ffd2.
```
$ uname -a
Linux localhost 4.4.78-perf+ #1 SMP PREEMPT Wed Nov 21 04:00:13 CST 2018 aarch64 Android
```
```
$ ./cmus --plugins
Input Plugins: /data/data/com.termux/files/usr/lib/cmus/ip
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg opus shn tak tta wav webm wma wv
    MIME Types:

Output Plugins: /data/data/com.termux/files/usr/lib/cmus/op
  pulse
```
```
$ ffmpeg -version
ffmpeg version 4.1.3 Copyright (c) 2000-2019 the FFmpeg developers
built with Android (5058415 based on r339409) clang version 8.0.2 (https://android.googlesource.com/toolchain/clang 40173bab62ec746213857d083c0e8b0abb568790) (https://android.googlesource.com/toolchain/llvm 7a6618d69e7e8111e1d49dc9e7813767c5ca756a) (based on LLVM 8.0.2svn)
configuration: --arch=aarch64 --as=aarch64-linux-android-clang --cc=aarch64-linux-android-clang --cxx=aarch64-linux-android-clang++ --cross-prefix=aarch64-linux-android- --disable-indevs --disable-outdevs --enable-indev=lavfi --disable-static --disable-symver --enable-cross-compile --enable-gnutls --enable-gpl --enable-libmp3lame --enable-libfreetype --enable-libvorbis --enable-libopus --enable-libx264 --enable-libx265 --enable-libxvid --enable-libvpx --enable-shared --enable-libsoxr --prefix=/data/data/com.termux/files/usr --target-os=android --extra-libs=-landroid-glob --enable-neon
libavutil      56. 22.100 / 56. 22.100
libavcodec     58. 35.100 / 58. 35.100
libavformat    58. 20.100 / 58. 20.100
libavdevice    58.  5.100 / 58.  5.100
libavfilter     7. 40.101 /  7. 40.101
libswscale      5.  3.100 /  5.  3.100
libswresample   3.  3.100 /  3.  3.100
libpostproc    55.  3.100 / 55.  3.100
```
I specifically removed all the input plugins except `cue` and `ffmpeg` just to be sure. The player freezes when I attempt to play any .ape file, no matter whether a .cue file exists next to it or not. I can still listen to flac, opus and m4a (and probably other, as I didn't test all the extensions reportedly supported by the input plugin) files using the `ffmpeg` plugin.
I also built cmus with debug logging enabled, but the log didn't contain any relevant info at the time of the freeze:
```
main: charset = 'UTF-8'
player_init: using real-time scheduling
player_init: setting priority to 99
player_init: could not create thread using real-time scheduling: Operation not permitted
player_set_op: setting op to 'pulse'
init_plugin: initialized mixer for pulse
init_plugin: initialized pcm for pulse
clear_pipe: read from pipe failed. errno = 11
init_curses: Number of supported colors: 256
init_curses: ts: 0 fs: 0
TIMER: worker job:           0.000166
TIMER: worker job:           0.007027
file_changed: file: /data/data/com.termux/files/home/storage/external-1/music/flac/album.flac
ip_setup: pcm convert: scale=1 convert=0 convert_in_place=0
_pa_context_running_cb: pulse: context state has changed to PA_CONTEXT_CONNECTING
_pa_context_running_cb: pulse: context state has changed to PA_CONTEXT_AUTHORIZING
_pa_context_running_cb: pulse: context state has changed to PA_CONTEXT_SETTING_NAME
_pa_context_running_cb: pulse: context state has changed to PA_CONTEXT_READY
_pa_stream_running_cb: pulse: stream state has changed to PA_STREAM_CREATING
_pa_stream_running_cb: pulse: stream state has changed to PA_STREAM_READY
main_loop: vol changed
file_changed: file: /data/data/com.termux/files/home/storage/external-1/music/ape/album.ape
ip_setup: pcm convert: scale=1 convert=0 convert_in_place=0
```

**Comment by shoober420** [comment]
I just tried playing back APE using deadbeef and ffmpeg, with the same results. This seems to be an issue with ffmpeg itself. I tried making a big on the ffmpeg tracker, but couldn’t create an account. I’ll try tomorrow.

**Comment by gdrbyKo1** [comment]
The freeze occurs here: https://github.com/cmus/cmus/blob/master/ip/ffmpeg.c#L306
The loop never finishes.

**Comment by gdrbyKo1** [comment]
The culprit lies in this `avcodec_send_packet` call in https://github.com/cmus/cmus/blob/master/ip/ffmpeg.c#L338:
```c
if (avcodec_send_packet(cc, &avpkt) == 0) {
	got_frame = !avcodec_receive_frame(cc, frame);
	if (got_frame)
		len = input->curr_pkt_size;
	else
		len = 0;
} else {
	got_frame = 0;
	len = 0;
}
```
After a few calls, it begins to return `AVERROR(EAGAIN)`, because there's still some data decoded from the previous packet which wasn't consumed yet, but the input plugin attempts to feed the decoder with another packet. Apparently, Monkey's Audio packets can decode into more than a single frame. The following `avcodec_receive_frame` should therefore be called more than once per an `avcodec_send_packet` call.

FFmpeg's documentation (https://www.ffmpeg.org/doxygen/3.4/group__lavc__encdec.html) says:
```
The API is very similar for encoding/decoding and audio/video, and works as follows:

    - Set up and open the AVCodecContext as usual.
    - Send valid input:
        - For decoding, call avcodec_send_packet() to give the decoder raw compressed data in an AVPacket.
        [...]
    - Receive output in a loop. Periodically call one of the avcodec_receive_*() functions and process their output:
        - For decoding, call avcodec_receive_frame(). On success, it will return an AVFrame containing uncompressed audio or video data.
        [...]
```
But the ffmpeg input plugin of cmus calls `avcodec_receive_frame` only once per packet and then continues with what follows in the audio file.

**Comment by shoober420** [comment]
I found the time to create a ticket for this issue on the ffmpeg bug tracker. I linked to this page so they can see all the detailed information you provided.

Thanks for doing all that debugging @gdrbyKo1

**Comment by shoober420** [comment]
So upon creating the ticket, it was just closed for being a duplicate. At least it’s a known bug.

https://trac.ffmpeg.org/ticket/7936#comment:1

**Comment by nefthy** [comment]
@shoober420 this actualy is a cmus bug not an ffmpeg one

@gdrbyKo1 would you like to make a PR?

**Comment by shoober420** [comment]
@nefthy Hmm, then this is a Deadbeef big as well. Very weird.

**Comment by gdrbyKo1** [comment]
@nefthy I don't think I'm up to the task, sorry.

**Comment by shoober420** [comment]
I tested ape playback directly through ffplay, and it works fine. Please @nefthy, consider fixing this bug. I have some Taake that needs blasted! lolol

**Comment by shoober420** [comment]
I thought I should mention a work around you can do if you have ffmpeg installed. You can convert ape to flac with ffmpeg. The command would be as follows.

ffmpeg -i poop.ape poop.flac

You can probably do wav files too.

**Comment by cpanceac** [comment]
APE playback (through .cue) still freezes cmus 2.9.1.

**Comment by timsofteng** [comment]
> APE playback (through .cue) still freezes cmus 2.9.1.

Same here.

**Comment by mightyrabbit99** [comment]
Same

> APE playback (through .cue) still freezes cmus 2.9.1.

**Comment by cpanceac** [comment]
Why is this closed? Is there a version newer than 2.9.1 where the problem is fixed?

**Comment by gavtroy** [comment]
Yes there is a newer version, 2.10.0 :grin: #1174 says it's fixed in that version.
